src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -0,0 +1,630 @@
+            {
+                // Avoid overwriting if the user already added it before
+                if (!extendedAttributes.ContainsKey(key))
+                {
+                    double unixTimeSeconds = ((double)(value.UtcDateTime - DateTime.UnixEpoch).Ticks)/TimeSpan.TicksPerSecond;
+                    extendedAttributes.Add(key, unixTimeSeconds.ToString("F6")); // 6 decimals, no commas
+                }
+            }
+
+            // Adds the specified string to the dictionary if it's longer than the specified max byte length.
+            static void TryAddStringField(Dictionary<string, string> extendedAttributes, string key, string value, int maxLength)
+            {
+                if (Encoding.UTF8.GetByteCount(value) > maxLength)
+                {
+                    extendedAttributes.Add(key, value);
+                }
+            }
+        }
+
+        // Generates an extended attribute key value pair string saved into a byte array, following the ISO/IEC 10646-1:2000 standard UTF-8 encoding format.
+        // https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html
+        private static byte[] GenerateExtendedAttributeKeyValuePairAsByteArray(byte[] keyBytes, byte[] valueBytes)
+        {
+            // Assuming key="ab" and value="cdef"
+
+            // The " ab=cdef\n" attribute string has a length of 9 chars
+            int suffixByteCount = 3 + // leading space, equals sign and trailing newline
+                keyBytes.Length + valueBytes.Length;
+
+            // The count string "9" has a length of 1 char
+            string suffixByteCountString = suffixByteCount.ToString();
+            int firstTotalByteCount = Encoding.ASCII.GetByteCount(suffixByteCountString);
+
+            // If we prepend the count string length to the attribute string,
+            // the total length increases to 10, which has one more digit
+            // "9 abc=def\n"
+            int firstPrefixAndSuffixByteCount = firstTotalByteCount + suffixByteCount;
+
+            // The new count string "10" has an increased length of 2 chars
+            string prefixAndSuffixByteCountString = firstPrefixAndSuffixByteCount.ToString();
+            int realTotalCharCount = Encoding.ASCII.GetByteCount(prefixAndSuffixByteCountString);
+
+            byte[] finalTotalCharCountBytes = Encoding.ASCII.GetBytes(prefixAndSuffixByteCountString);
+
+            // The final string should contain the correct total length now
+            List<byte> bytesList = new();
+
+            bytesList.AddRange(finalTotalCharCountBytes);
+            bytesList.Add(TarHelpers.SpaceChar);
+            bytesList.AddRange(keyBytes);
+            bytesList.Add(TarHelpers.EqualsChar);
+            bytesList.AddRange(valueBytes);
+            bytesList.Add(TarHelpers.NewLineChar);
+
+            Debug.Assert(bytesList.Count == (realTotalCharCount + suffixByteCount));
+
+            return bytesList.ToArray();
```

Comment:
This is allocating a byte[] for finalTotalCharCountBytes, potentially multiple byte[]s backing the list, the list itself, and the final byte[] from ToArray().  You've already computed how big it needs to be... you can just allocate the final array and populate it.